nbgrader/docs/source/user_guide/what_is_nbgrader.rst

```diff
@@ -0,0 +1,350 @@
+work (this has been a problem with a few languages kernels in the
+past).
+
+TODO: partial credit.  If a autograder test cell outputs a single
+number to the standard output stream, then it will use that as the
+number of points.  However, this could always be simulated by dividing
+the autograded task into multiple cells.
+
+
+
+Validation
+~~~~~~~~~~
+
+**Validation** is very related to autograding.  There is a button on
+the student interface marked "validate", which executes the student
+version of the notebook from top to bottom, and reports any errors.
+This is exactly equivalent to "Restart and run all", but doesn't stop
+on errors.  Since all it can access is the actual notebook file the
+student has, it can not take into account the hidden tests.  If an
+instructor wants a test to be visible to the students.
+
+There is currently no support for inserting hidden tests into the
+notebook file (perhaps you could in a hidden cell, but since the
+student actually has the file... it's not going to be hidden to anyone
+willing to do a bit of exploration).
+
+
+
+Manual grading
+--------------
+
+After autograding, there is a web UI to do **manual grading**.  This
+allows one to see the output from autograding, give comments, adjust
+points, etc.  There are also purely manually graded exercises.
+
+The output from manual grading is only stored in ``gradebook.db``, and
+is merged into the final output at the ``feedback`` step.
+
+
+
+gradebook.db and student management
+-----------------------------------
+
+The **gradebook** or **database** is stored (by default) at
+``gradebook.db`` at the root of the course directory.  Out of the box,
+it is sqlite3, but can be other database systems, too.
+
+First, the gradebook stores student mappings.  It stores a
+``student_id`` (string) that is the name used on the filesystem for
+each student.  It can also store a firstname/lastname/email for each
+student, but it doesn't try to replace a complex student management
+system.
+
+The database also stores assignments and their cells.  For example, it
+stores the contents of read-only cells, and autograder tests cells,
+which get re-inserted into the notebook before the autograde step.
+
+Grades can be exported in csv format.  You can also build other
+exporters, which access the database and export somehow - to a file,
+or perhaps other fancy things like uploading directly.
+
+
+
+Feedback
+--------
+
+Feedback is a HTML file, basically like a rendering of the notebook
+using nbconvert.  However, it adds in points and feedback.
+
+Historically, feedback was just generated, and it was up to the
+teacher to distribute it somehow (for example, uploading to the course
+management system or scripting copying it into users home
+directories).  Now, using the exchange, there can be automatically
+distributed.  This is described above.
+
+
+Web extensions
+--------------
+
+TODO
+
+
+See also
+--------
+
+TODO
```

Comment:
I will shortly have some extensive documentation for you to copy/refer to - which we've written for our service.

This documentation will be public-access.

Reply:
Sorry - forgot about this.

These may be of interest: https://noteable.edina.ac.uk/student_guide/ & https://noteable.edina.ac.uk/nbguide/ may help